### PR TITLE
[+] CORE : Bring back "enable / disable on mobile" action

### PR DESCRIPTION
**Pull request by Quetzacoalt91**
## Description

This pull-request brings back the mobile option on PrestaShop.
capture du 2016-04-06 18-03-03
## Steps to Test this Fix

In order to test this pull-request:
- Go to your module catalog, tab "Installed modules" or "Notifications".
- Open the contextual menu of an **Installed** module
- Check that "Enabled mobile" or "Disable mobile" appear in the list
- Click on it and see an ajax request executed
- Everything should be went right and the other option should now appear
## Forge Ticket (optional)

This pull-request fixes the issue [BOOM-74](http://forge.prestashop.com/browse/BOOM-74)

### CODE REVIEW

**Comment by vTerenti** [comment]
QA OK

File: src/Adapter/Module/ModuleDataProvider.php [hunk 1]
```diff
@@ -133,4 +135,25 @@ protected function getModulesDir()
...
+
+    /**
+     * Check if the module has been enabled on mobile
+     * @param string $name The technical module name to check
+     * @return int The devices enabled for this module
+     */
+    private function getDeviceStatus($name)
+    {
+        $id_shops = (new Context())->getContextListShopID();
+        // ToDo: Load list of all installed modules ?
```
Comment (mickaelandrieu) [inline src/Adapter/Module/ModuleDataProvider.php]: we may create a ticket on forge instead :)

https://www.youtube.com/watch?v=ZXsQAXx_ao0

File: src/Adapter/Module/ModulePresenter.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
+<?php
+
```
Comment (mickaelandrieu) [inline src/Adapter/Module/ModulePresenter.php]: header class please

File: src/Adapter/Module/ModulePresenter.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+     */
+    private $currency;
+
+    /** @var PriceFormatter */
+    private $priceFormatter;
+
+    public function __construct(Currency $currency)
+    {
+        $this->currency = $currency;
+        // Not declared as a Symfony service :(
```
Comment (mickaelandrieu) [inline src/Adapter/Module/ModulePresenter.php]: hmm .. ? declare it then ^^

File: src/Core/Addon/Module/ModuleInterface.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+interface ModuleInterface extends AddonInterface
+{
+    public function onInstall();
+    public function onUninstall();
+    public function onEnable();
+    public function onDisable();
+    public function onMobileEnable();
+    public function onMobileDisable();
+    public function onReset();
+
```
Comment (mickaelandrieu) [inline src/Core/Addon/Module/ModuleInterface.php]: remove empty line please :)

File: src/Core/Addon/Module/ModuleManager.php [hunk 5]
```diff
@@ -224,6 +224,48 @@ public function enable($name)
     }
 
     /**
+     * Disable a module specifiquely on mobile.
```
Comment (mickaelandrieu) [inline src/Core/Addon/Module/ModuleManager.php]: specifically ?

File: src/Core/Addon/Module/ModuleManager.php [hunk 6]
```diff
@@ -224,6 +224,48 @@ public function enable($name)
...
+     *
+     * @param  string $name The module name to disable
+     * @return bool         True for success
+     */
+    public function disable_mobile($name)
```
Comment (mickaelandrieu) [inline src/Core/Addon/Module/ModuleManager.php]: camelCase notation please (for consistency with `isEnabled`), same for `enable_mobile`

File: src/Core/Addon/Module/ModuleManager.php [hunk 7]
```diff
@@ -224,6 +224,48 @@ public function enable($name)
...
+    {
+        if (!$this->employee->can('edit', 'AdminModules')
+            || !$this->moduleProvider->can('configure', $name)) {
+            throw new Exception('You are not allowed to disable this module on mobile');
```
Comment (mickaelandrieu) [inline src/Core/Addon/Module/ModuleManager.php]: `You are not allowed to disable this module on mobile.`

File: src/PrestaShopBundle/Resources/views/Admin/Module/Includes/action_menu.html.twig [hunk 8]
```diff
@@ -1,25 +1,25 @@
 <div class="pull-right btn-group">
 
-    {% if module.attributes.get('url_active') == 'buy' %}
-        <a class="btn btn-primary btn-sm light-button module_action_menu_go_to_addons" href="{{module.attributes.get('url') }}" target="_blank">
-            {% if module.attributes.has('price') and module.attributes.get('price').EUR > 0 %}{{module.attributes.get('price').EUR}}{% else %}Discover{% endif %}
+    {% if module.attributes.url_active == 'buy' %}
+        <a class="btn btn-primary btn-sm light-button module_action_menu_go_to_addons" href="{{module.attributes.url }}" target="_blank">
+            {% if module.attributes.price.raw != '0.00' %}{{module.attributes.price.displayPrice}}{% else %}Discover{% endif %}
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Resources/views/Admin/Module/Includes/action_menu.html.twig]: who have introduced this option ? the main idea of presenter is to avoid any calculation operations on templates :disappointed:

File: src/Core/Addon/Module/ModuleManager.php [hunk 9]
```diff
@@ -224,6 +224,48 @@ public function enable($name)
...
+        }
+    }
+
+    /**
+     * Enable a module previously disabled on mobile
+     *
+     * @param  string $name The module name to enable
+     * @return bool         True for success
+     */
+    public function enable_mobile($name)
```
Comment (mickaelandrieu) [inline src/Core/Addon/Module/ModuleManager.php]: Can you add unit tests ? now we have entirely covered theses classes, let's not give up ^^

File: tests/Unit/Core/Addon/Module/ModuleRepositoryTest.php [hunk 10]
```diff
@@ -194,7 +194,7 @@ public function test_get_not_enabled_modules()
 
         foreach ($all_modules as $name => $module) {
             // Each installed module must be found in the installed modules list
-            if ($module->attributes->get('productType') == 'module' && $module->database->has('active') && $module->database->get('active') == 0) {
+            if ($module->attributes->get('productType') == 'module' && $module->database->get('installed') == 1  && $module->database->get('active') == 0) {
```
Comment (mickaelandrieu) [inline tests/Unit/Core/Addon/Module/ModuleRepositoryTest.php]: hmm this smell bad. If this test pass before, it should pass after without any update.

Can you explain me why you need to check for `activate` field now ?

File: src/PrestaShopBundle/Resources/views/Admin/Module/Includes/action_menu.html.twig (hunk 8 above)
Comment (Quetzacoalt91) [inline src/PrestaShopBundle/Resources/views/Admin/Module/Includes/action_menu.html.twig]: This field allow us to know a product is free without having to deal with the currency (Generated by the presenter)

File: tests/Unit/Core/Addon/Module/ModuleRepositoryTest.php (hunk 10 above)
Comment (Quetzacoalt91) [inline tests/Unit/Core/Addon/Module/ModuleRepositoryTest.php]: With this PR, we do not have to test if a attribute exists anymore. They always have (at least) a default value, so the check is now different.

File: src/Adapter/Module/ModulePresenter.php (hunk 2 above)
Comment (Quetzacoalt91) [inline src/Adapter/Module/ModulePresenter.php]: Will be done, but it appears that another classes do not have their header too (this file comes from a copy-paste)

File: src/Adapter/Module/ModulePresenter.php (hunk 3 above)
Comment (Quetzacoalt91) [inline src/Adapter/Module/ModulePresenter.php]: Will be done for the ModulePresenter, but the other presenters will be declared by the guy who created them ;)

File: tests/Unit/Core/Addon/Module/ModuleRepositoryTest.php (hunk 10 above)
Comment (mickaelandrieu) [inline tests/Unit/Core/Addon/Module/ModuleRepositoryTest.php]: ok but why did you add a check on the `installed` attribute? I mean, if the module is active, it must be installed, right ?

Comment (Quetzacoalt91) [inline tests/Unit/Core/Addon/Module/ModuleRepositoryTest.php]: You're right, but here we have a module which is not `active`. And both modules installed and uninstalled can be not active. :wink:

File: src/Core/Addon/Module/ModuleManager.php (hunk 9 above)
Comment (Quetzacoalt91) [inline src/Core/Addon/Module/ModuleManager.php]: Done

**Comment by mickaelandrieu** [comment]
you can rebase from develop, tests are fiuxed
